src/coreclr/jit/lower.cpp

```diff
@@ -3504,8 +3504,10 @@ GenTree* Lowering::LowerTailCallViaJitHelper(GenTreeCall* call, GenTree* callTar
     assert(argEntry != nullptr);
     GenTree* arg0 = argEntry->GetEarlyNode()->AsPutArgStk()->gtGetOp1();
 
+    // Temporarily link in the call target range just before the call, since it
+    // may refer to a spilled temp. We will move the putarg after this via MoveCFGCallArgs below.
     ContainCheckRange(callTargetRange);
-    BlockRange().InsertAfter(arg0, std::move(callTargetRange));
+    BlockRange().InsertBefore(call, std::move(callTargetRange));
```

Comment:
I think this is right since it matches how we normally insert the call target. However, I am unsure what the code above that just removes `gtCallAddr` is doing. That code seems odd. Do you have any insight into it?

From reading the code I am guessing we have two separate cases: one where `callTarget` is `call->gtCallAddr` and thus is already inserted into LIR, and another where `callTarget` was just created and hasn't been inserted yet.

I think for the former case the handling here is wrong -- we should not be moving the call target IR around.

I would suggest inserting `callTargetRange` before the call to `LowerTailCallViaJitHelper` in `LowerCall`, when necessary, and updating `LowerTailCallViaJitHelper` to stop trying to move those nodes around.

Reply:
Reminds me that at some point I'd like to go rework how GT_CALL handles all this and not have two seemingly similar mechanisms for specifying where control should go.

I'll try inserting earlier.